File: ts-sdk/client/src/txbuilder/closeTunaSpotPosition.ts
import {
  type Account,
  Address,
  GetAccountInfoApi,
  GetMultipleAccountsApi,
  IInstruction,
  Rpc,
  TransactionSigner,
} from "@solana/kit";
import { fetchAllMaybeMint, findAssociatedTokenPda, Mint } from "@solana-program/token-2022";
import assert from "assert";

import { fetchMaybeTunaSpotPosition, getCloseTunaSpotPositionInstruction } from "../generated";
import { getMarketAddress, getTunaSpotPositionAddress } from "../pda.ts";

export async function closeTunaSpotPositionInstructions(
  rpc: Rpc<GetAccountInfoApi & GetMultipleAccountsApi>,
  authority: TransactionSigner,
  pool: Address,
): Promise<IInstruction[]> {
  const tunaPositionAddress = (await getTunaSpotPositionAddress(authority.address, pool))[0];
  const tunaPosition = await fetchMaybeTunaSpotPosition(rpc, tunaPositionAddress);
  if (!tunaPosition.exists) throw new Error("Tuna position account not found");

  const [mintA, mintB] = await fetchAllMaybeMint(rpc, [tunaPosition.data.mintA, tunaPosition.data.mintB]);
  assert(mintA.exists, "Token A account not found");
  assert(mintB.exists, "Token B account not found");

  const ix = await closeTunaSpotPositionInstruction(authority, pool, mintA, mintB);
  return [ix];
}

export async function closeTunaSpotPositionInstruction(
  authority: TransactionSigner,
  pool: Address,
  mintA: Account<Mint>,
  mintB: Account<Mint>,
): Promise<IInstruction> {
  const tunaPositionAddress = (await getTunaSpotPositionAddress(authority.address, pool))[0];
  const marketAddress = (await getMarketAddress(pool))[0];

  const tunaPositionAtaA = (
    await findAssociatedTokenPda({
      owner: tunaPositionAddress,
      mint: mintA.address,
      tokenProgram: mintA.programAddress,
    })
  )[0];

  const tunaPositionAtaB = (
    await findAssociatedTokenPda({
      owner: tunaPositionAddress,
      mint: mintB.address,
      tokenProgram: mintB.programAddress,
    })
  )[0];

  return getCloseTunaSpotPositionInstruction({
    mintA: mintA.address,
    mintB: mintB.address,
    tokenProgramA: mintA.programAddress,
    tokenProgramB: mintB.programAddress,
    market: marketAddress,
    tunaPositionAtaA,
    tunaPositionAtaB,
    authority,
    tunaPosition: tunaPositionAddress,
  });
}

File: rust-sdk/client/src/txbuilder/close_tuna_spot_position.rs
use crate::accounts::{fetch_tuna_spot_position, TunaSpotPosition};
use crate::instructions::CloseTunaSpotPosition;
use crate::{get_market_address, get_tuna_spot_position_address};
use anyhow::{anyhow, Result};
use solana_client::rpc_client::RpcClient;
use solana_instruction::Instruction;
use solana_pubkey::Pubkey;
use spl_associated_token_account::get_associated_token_address_with_program_id;

pub fn close_tuna_spot_position_instructions(rpc: &RpcClient, authority: &Pubkey, pool: &Pubkey) -> Result<Vec<Instruction>> {
    let tuna_position_address = get_tuna_spot_position_address(authority, pool).0;
    let tuna_position = fetch_tuna_spot_position(rpc, &tuna_position_address)?;

    let mint_a_address = tuna_position.data.mint_a;
    let mint_b_address = tuna_position.data.mint_b;

    let mint_accounts = rpc.get_multiple_accounts(&[mint_a_address.into(), mint_b_address.into()])?;
    let mint_a_account = mint_accounts[0].as_ref().ok_or(anyhow!("Token A mint account not found"))?;
    let mint_b_account = mint_accounts[1].as_ref().ok_or(anyhow!("Token B mint account not found"))?;

    Ok(vec![close_tuna_spot_position_instruction(
        authority,
        &tuna_position.data,
        &tuna_position.address,
        &mint_a_account.owner,
        &mint_b_account.owner,
    )])
}

pub fn close_tuna_spot_position_instruction(
    authority: &Pubkey,
    tuna_position: &TunaSpotPosition,
    tuna_position_address: &Pubkey,
    token_program_a: &Pubkey,
    token_program_b: &Pubkey,
) -> Instruction {
    let market_address = get_market_address(&tuna_position.pool).0;

    let ix_builder = CloseTunaSpotPosition {
        authority: *authority,
        mint_a: tuna_position.mint_a,
        mint_b: tuna_position.mint_b,
        token_program_a: *token_program_a,
        token_program_b: *token_program_b,
        market: market_address,
        tuna_position: *tuna_position_address,
        tuna_position_ata_a: get_associated_token_address_with_program_id(tuna_position_address, &tuna_position.mint_a, token_program_a),
        tuna_position_ata_b: get_associated_token_address_with_program_id(tuna_position_address, &tuna_position.mint_b, token_program_b),
    };

    ix_builder.instruction()
}
